qiskit_experiments/analysis/curve_analysis.py

```diff
@@ -0,0 +1,681 @@
+        xdata: a 1D float array of x-data.
+        ydata: a 1D float array of y-data.
+        p0: initial guess for optimization parameters.
+        sigma: Optional, a 1D array of standard deviations in ydata in absolute units.
+        bounds: Optional, lower and upper bounds for optimization parameters.
+        kwargs: additional kwargs for scipy.optimize.curve_fit.
+
+    Returns:
+        result containing ``popt`` the optimal fit parameters,
+        ``popt_err`` the standard error estimates popt,
+        ``pcov`` the covariance matrix for the fit,
+        ``reduced_chisq`` the reduced chi-squared parameter of fit,
+        ``dof`` the degrees of freedom of the fit,
+        ``xrange`` the range of xdata values used for fit.
+
+    Raises:
+        QiskitError: if the number of degrees of freedom of the fit is
+                     less than 1.
+
+    .. note::
+        ``sigma`` is assumed to be specified in the same units as ``ydata``
+        (absolute units). If sigma is instead specified in relative units
+        the `absolute_sigma=False` kwarg of scipy curve_fit must be used.
+        This affects the returned covariance ``pcov`` and error ``popt_err``
+        parameters via ``pcov(absolute_sigma=False) = pcov * reduced_chisq``
+        ``popt_err(absolute_sigma=False) = popt_err * sqrt(reduced_chisq)``.
+    """
+    # Check the degrees of freedom is greater than 0
+    dof = len(ydata) - len(p0)
+    if dof < 1:
+        raise QiskitError(
+            "The number of degrees of freedom of the fit data and model "
+            " (len(ydata) - len(p0)) is less than 1"
+        )
+
+    # Override scipy.curve_fit default for absolute_sigma=True
+    # if sigma is specified.
+    if sigma is not None and "absolute_sigma" not in kwargs:
+        kwargs["absolute_sigma"] = True
+
+    # Run curve fit
+    # pylint: disable = unbalanced-tuple-unpacking
+    popt, pcov = opt.curve_fit(
+        f=func, xdata=xdata, ydata=ydata, sigma=sigma, p0=p0, bounds=bounds, **kwargs
+    )
+    popt_err = np.sqrt(np.diag(pcov))
+
+    # Calculate the reduced chi-squared for fit
+    yfits = func(xdata, *popt)
+    residues = (yfits - ydata) ** 2
+    if sigma is not None:
+        residues = residues / (sigma ** 2)
+    reduced_chisq = np.sum(residues) / dof
+
+    # Compute xdata range for fit
+    xdata_range = [min(xdata), max(xdata)]
+
+    result = {
+        "popt": popt,
+        "popt_err": popt_err,
+        "pcov": pcov,
+        "reduced_chisq": reduced_chisq,
+        "dof": dof,
+        "xrange": xdata_range,
+    }
+
+    return AnalysisResult(result)
+
+
+class CurveAnalysis(BaseAnalysis):
+    """A base class for curve fit type analysis.
+
+    The subclasses can override class attributes to define the behavior of
+    data extraction and fitting. This docstring describes how code developers can
+    create a new curve fit analysis subclass inheriting from this base class.
+
+    Class Attributes:
```

Comment:
I first added default data processor to the class attributes, but I was suffering from passing runtime options that may be needed to initialize the processor. Now I realized we can solve this issue by passing a calibrated processor instance itself as a runtime option. I also added `__default_data_processor__` as a class attribute, so that we can avoid passing data processor for every run, i.e. some data processing chain doesn't need calibration, and we can use these processor without calibration -- thus can be hardcoded in the attribute.

Reply:
Anyways I added `_calibrate_data_processor`. This is called only when you use DataProcessor instance. I don't think `__cal_key__` works, because we have several types of training. Off the top of my head:

1. SVD for level1 data
This doesn't need separate experiment. Take all measurement data and calculate a principle component to project scatters to the single axis.

2. Discriminator
This requires separate experiment. Thus we need something like `__cal_key__`. This also requires extra information of labels of the prepared states.

3. Custom mitigation
Mitigation can be supported on provide end, but user may want to test custom mitigation algorithm. This can be registered as a processor node and calibrated with the test experiment. This is similar situation with 2 but for level2 data.

I think developer should be able to write calibration sequence more flexibly. It seems like providing a processor calibration as an overridable method is the best approach for now.